myst_nbs/howto/howto_debugging.myst.md

```diff
@@ -6,108 +6,184 @@ jupytext:
-    trace = pm.sample(
-        5, step, tune=0, chains=1, progressbar=False
-    )  # Make sure not to draw too many samples
+    trace = pm.sample(5, step, tune=0, chains=1, progressbar=False)
+```
 
-sys.stdout = old_stdout
+Exception handling of PyMC v4 has improved, so now SamplingError exception prints out the intermediate values of `mu` and `sd` which led to likelihood of `-inf`. However, this technique of printing intermediate values with `aeasara.printing.Print` can be valuable in more complicated cases.
 
-output = mystdout.getvalue().split("\n")
-mulines = [s for s in output if "mu" in s]
++++
+
+### Bringing it all together
+
+```{code-cell} ipython3
+rng = np.random.default_rng(RANDOM_SEED)
+y = rng.normal(loc=5, size=20)
+
+old_stdout = sys.stdout
+mystdout = sys.stdout = StringIO()
+
+with pm.Model() as model:
+    mu = pm.Normal("mu", mu=0, sigma=10)
+    a = pm.Normal("a", mu=0, sigma=10, initval=0.1)
+    b = pm.Normal("b", mu=0, sigma=10, initval=0.1)
+    sd_print = Print("Delta")(a / b)
+    obs = pm.Normal("obs", mu=mu, sigma=sd_print, observed=y)
+
+    # limiting number of samples and chains to simplify output
+    trace = pm.sample(draws=10, tune=0, chains=1, progressbar=False)
 
-muvals = [line.split()[-1] for line in mulines]
-plt.plot(np.arange(0, len(muvals)), muvals)
-plt.xlabel("proposal iteration")
-plt.ylabel("mu value");
+output = mystdout.getvalue()
+sys.stdout = old_stdout  # setting sys.stdout back
 ```
 
 ```{code-cell} ipython3
-trace["mu"]
+output
 ```
 
-Notice that for each iteration, 3 values were printed and recorded. The printed values are the original value (last sample), the proposed value and the accepted value. Plus the starting value in the very beginning, we recorded in total `1+3*5=16` value above.
+Raw output is a bit messy and requires some cleanup and formatting to convert to `numpy.ndarray`. In the example below regex is used to clean up the output, and then it is evaluated with `eval` to give a list of floats. Code below also works with higher-dimensional outputs (in case you want to experiment with different models).
+
+```{code-cell} ipython3
+import re
+
+# output cleanup and conversion to numpy array
+# this is code accepts more complicated inputs
+pattern = re.compile("Delta __str__ = ")
+output = re.sub(pattern, " ", output)
+pattern = re.compile("\\s+")
+output = re.sub(pattern, ",", output)
+pattern = re.compile(r"\[,")
+output = re.sub(pattern, "[", output)
+output += "]"
+output = "[" + output[1:]
+output = eval(output)
+output = np.array(output)
+```
+
+```{code-cell} ipython3
+output
+```
+
+Notice that we requested 5 draws, but got 34 sets of $a/b$ values. The reason is that for each iteration, all proposed values are printed (not just the accepted values). Negative values are clearly problematic.
+
+```{code-cell} ipython3
+output.shape
+```
 
 ```{code-cell} ipython3
 %load_ext watermark
-%watermark -n -u -v -iv -w
+%watermark -n -u -v -iv -w -p aesara,xarray,patsy
 ```
```

Comment:
There should be a markdown cell with the `## Watermark` title right before the watermark code cell so it is accessible from the page table of contents (right sidebar). https://docs.pymc.io/en/latest/contributing/jupyter_style.html#watermark.

Above that there should also be the authorship section: https://docs.pymc.io/en/latest/contributing/jupyter_style.html#authorship-and-attribution. You don't need to add all the info in the template nor track all the history of the notebook. But it needs to include all the already known and available info. That currently is Thomas authored the notebook and you updated it (here we also know the date and PR).